### PR TITLE
Refactor GetBase heuristics for kernels.

### PR DESCRIPTION
**Pull request by s-kanev**
Not a functional change. Just separate a helper function to match most kernels heuristics.

This way they can be shared between ET_EXEC and ET_DYN kernels.

Tested: unit

### CODE REVIEW

**Review by gmarin13 — DISMISSED** [review]

File: internal/elfexec/elfexec.go [hunk 1]
```diff
@@ -165,19 +165,61 @@ func GetBuildID(binary io.ReaderAt) ([]byte, error) {
...
-// address to get symbol table address. For an executable, the base
-// is 0. Otherwise, it's a shared library, and the base is the
-// address where the mapping starts. The kernel is special, and may
-// use the address of the _stext symbol as the mmap start. _stext
-// offset can be obtained with `nm vmlinux | grep _stext`
-func GetBase(fh *elf.FileHeader, loadSegment *elf.ProgHeader, stextOffset *uint64, start, limit, offset uint64) (uint64, error) {
+// kernelBase caluclates the base for kernel mappings, which usually require
+// special handling. For kernel mappings, tools (like perf) use the address of
+// the kernel relocation symbol (_text or _stext) as the mmap start. Additionaly,
+// for obfuscation, ChromeOS kernels can get remapped to the 0-th page.
```
[the remark this answers is not in the document]
Comment (s-kanev) [reply internal/elfexec/elfexec.go]: Done, thanks.

File: internal/elfexec/elfexec.go [hunk 2]
```diff
@@ -202,43 +244,15 @@ func GetBase(fh *elf.FileHeader, loadSegment *elf.ProgHeader, stextOffset *uint6
...
-				// adjust for this before calling GetBase, in which case the page
-				// alignment should be different from that of stextOffset.
-				return start - *stextOffset, nil
-			}
-
+		// ChromeOS can remap its kernel to 0, and the caller might have not found
+		// the _stext symbol. Split this case from kernelBase() above, since we don't
+		// want to apply it to an ET_DYN user-mode executable.
+		if start == 0 && limit != 0 && stextOffset == nil {
 			return start - loadSegment.Vaddr, nil
```
[the remark this answers is not in the document]
Comment (s-kanev) [reply internal/elfexec/elfexec.go:255]: Yes, that's intentional for readability. So it's obvious it's still the same calculation as other cases (with just some zeros), and not different custom formulas.

**Comment by s-kanev** [comment]
The CI issue is separate and not related to this CL. Can you merge (I don't have permissions)? Thanks.

**Review by gmarin13 — APPROVED** [review]

**Comment by gmarin13** [comment]
@nolanmar511 , do you know why some CI tests are failing and how to fix them? They seem unrelated to this PR.

**Comment by s-kanev** [comment]
> @nolanmar511 , do you know why some CI tests are failing and how to fix them? They seem unrelated to this PR.

FWIW, I filed #662 with what I found out about the failures, but ran out of knowledge in terms of a fixing.

**Comment by gmarin13** [comment]
> > @nolanmar511 , do you know why some CI tests are failing and how to fix them? They seem unrelated to this PR.
> 
> FWIW, I filed #662 with what I found out about the failures, but ran out of knowledge in terms of a fixing.

Thanks for filling. I cannot merge the PR unless all required statuses pass.

**Pull request merged**